### PR TITLE
feat: add traffic manager backend controller to handle delete

### PR DESCRIPTION
**Pull request by zhiying-lin**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:

- [ ] run `make reviewable` for basic local test
- [ ] Read and followed fleet-networking's [Code of conduct](https://github.com/Azure/fleet-networking/blob/main/CODE_OF_CONDUCT.md).
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests

### CODE REVIEW

File: pkg/controllers/hub/trafficmanagerbackend/controller.go [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+	// AzureResourceEndpointNameFormat is the name format of the Azure Traffic Manager Endpoint created by the fleet controller.
+	// The naming convention of a Traffic Manager Endpoint is fleet-{TrafficManagerBackendUUID}#{ServiceImportName}#{ClusterName}.
+	// All the object name length should be restricted to <= 63 characters.
+	// The endpoint name must contain no more than 260 characters, excluding the following characters "< > * % $ : \ ? + /".
+	AzureResourceEndpointNameFormat = AzureResourceEndpointNamePrefix + "%s#%s"
+)
+
+var (
+	// create the func as a variable so that the integration test can use a customized function.
+	generateAzureTrafficManagerProfileNameFunc = func(profile *fleetnetv1alpha1.TrafficManagerProfile) string {
```
[the remark this answers is not in the document]
Comment (zhiying-lin) [reply pkg/controllers/hub/trafficmanagerbackend/controller.go:53]: it's using profile UID and hard to construct beforehand in the IT

File: pkg/controllers/hub/trafficmanagerbackend/controller.go [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+	}
+	klog.V(2).InfoS("Removed trafficManagerBackend finalizer", "trafficManagerBackend", backendKObj)
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) deleteAzureTrafficManagerEndpoints(ctx context.Context, backend *fleetnetv1alpha1.TrafficManagerBackend) error {
+	backendKObj := klog.KObj(backend)
+	profile := &fleetnetv1alpha1.TrafficManagerProfile{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: backend.Spec.Profile.Name, Namespace: backend.Namespace}, profile); err != nil {
+		if apierrors.IsNotFound(err) {
```
[the remark this answers is not in the document]
Comment (zhiying-lin) [reply pkg/controllers/hub/trafficmanagerbackend/controller.go:140]: no matter whether we block or not, the user can manually delete the profile by removing the finalizer and delete the object, we still need to handle this case.

File: pkg/controllers/hub/trafficmanagerbackend/controller.go [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+		klog.V(2).InfoS("Azure Traffic Manager profile does not exist", "trafficManagerBackend", backendKObj, "trafficManagerProfile", profileKObj, "azureProfileName", azureProfileName)
+		return nil // skip handling endpoints deletion
+	}
+	if getRes.Profile.Properties == nil {
+		klog.V(2).InfoS("Azure Traffic Manager profile has nil properties and skipping handling endpoints deletion", "trafficManagerBackend", backendKObj, "trafficManagerProfile", profileKObj, "azureProfileName", azureProfileName)
+		return nil
+	}
+
+	klog.V(2).InfoS("Deleting Azure Traffic Manager endpoints", "trafficManagerBackend", backendKObj, "trafficManagerProfile", backend.Spec.Profile.Name)
+	errs, cctx := errgroup.WithContext(ctx)
```
[the remark this answers is not in the document]
Comment (zhiying-lin) [reply pkg/controllers/hub/trafficmanagerbackend/controller.go:165]: Yes, cause one serviceImport stands for multiple backends (svc), so we'll create multiple endpoints here.

File: pkg/controllers/hub/trafficmanagerbackend/controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jwtty) [reply pkg/controllers/hub/trafficmanagerbackend/controller.go:53]: Can we test it by getting the obj after creating in the test? I feel this is too hacky.

[the remark this answers is not in the document]
Comment (zhiying-lin) [reply pkg/controllers/hub/trafficmanagerbackend/controller.go:53]: we're using fake atm server, so cannot predict the name if using UID. We'll test the whole implementation using real atm server in the e2e tests.

**Review by jwtty — APPROVED** [review]

**Pull request merged**